ndd_slurm.py

```diff
@@ -184,42 +190,87 @@ def get_slaves(args):
         raise ValueError('one of -D or -d must be specified')
 
 
-def run_slave(args):
-    procs = []
+def run_slave_tar(args, procs):
+    read_fd, write_fd = os.pipe()
+    procs.append(init_process(['tar', '-xC', args.o], read_fd=read_fd))
+    return write_fd
+
+
+def run_slave_comp(args, procs):
+    read_fd, write_fd = os.pipe()
     if args.r:
-        read_fd, write_fd = os.pipe()
-        procs.append(init_process(['tar', '-xC', args.o], read_fd=read_fd))
+        tar_w = run_slave_tar(args, procs)
+        procs.append(init_process(['pigz', '-d'], read_fd=read_fd,
+                                  write_fd=tar_w))
     else:
+        out_fd = os.open(args.o, os.O_CREAT | os.O_WRONLY)
+        procs.append(init_process(['pigz', '-dc'], read_fd=read_fd,
+                                  write_fd=out_fd))
+    return write_fd
+
+
+def run_slave(args):
+    procs = []
+    if not (args.r or args.z):
         write_fd = None
+    if args.z:
+        write_fd = run_slave_comp(args, procs)
+    elif args.r:
+        write_fd = run_slave_tar(args, procs)
     if args.H:
-        slave = args.c
-        slaves = args.S.split(',')
-        cmd = get_ssh_slave_ndd_cmd(args, slave, slaves, write_fd)
+        cmd = get_ssh_slave_ndd_cmd(args, write_fd)
     else:
         cmd = get_slave_ndd_cmd(args, write_fd)
-    procs.append(init_process(cmd))
+    procs.append(init_process(cmd, write_fd=write_fd))
     return wait(procs)
 
 
-def run_master(args):
-    procs = []
+def run_master_tar(args, procs):
+    read_fd, write_fd = os.pipe()
+    procs.append(init_process(['tar', '-c', args.i],
+                              write_fd=write_fd))
+    return read_fd
+
+
+def run_master_comp(args, procs):
+    read_fd, write_fd = os.pipe()
     if args.r:
-        read_fd, write_fd = os.pipe()
-        procs.append(init_process(['tar', '-c', args.i], write_fd=write_fd))
+        tar_r = run_master_tar(args, procs)
+        procs.append(init_process(['pigz', '--fast'], read_fd=tar_r,
+                                  write_fd=write_fd))
     else:
+        in_fd = os.open(args.i, os.O_RDONLY)
+        procs.append(init_process(['pigz', '--fast', '-c'], read_fd=in_fd,
```

Comment:
Probably `-c` is not needed here: earlier in the same command we execute it without `-c` and go well with stdin as input and stdout as output.

Also after this you may eliminate double `init_process` calls, which differ only in one argument.